numpy/_core/src/multiarray/dtypemeta.c

```diff
@@ -1206,12 +1207,12 @@ dtypemeta_wrap_legacy_descriptor(PyArray_Descr *descr,
             dt_slots->default_descr = void_default_descr;
             dt_slots->discover_descr_from_pyobject = (
                     void_discover_descr_from_pyobject);
-            dt_slots->common_instance = void_common_instance;
-            dt_slots->ensure_canonical = void_ensure_canonical;
+            dt_slots->common_instance = (void *)void_common_instance;
+            dt_slots->ensure_canonical = (void *)void_ensure_canonical;
             dt_slots->get_fill_zero_loop =
-                    npy_get_zerofill_void_and_legacy_user_dtype_loop;
+                    (void *)npy_get_zerofill_void_and_legacy_user_dtype_loop;
             dt_slots->get_clear_loop =
-                    npy_get_clear_void_and_legacy_user_dtype_loop;
+                    (void *)npy_get_clear_void_and_legacy_user_dtype_loop;
```

Comment:
are these `void*` casts a remnant of an earlier version or this PR where you modified the `NPY_DType_Slots` struct or do you need these for some other reason?

Reply:
The void casts are a bit lazy (could use the full name, which is long though), but because of the changing the signature, you need to cast now to avoid compiler warnings.

Reply:
instead of adding these void * casts, I'd prefer adding new function typedefs for internal usages, even better if we can figure out how to make this function take the real struct

Reply:
Yeah, sorry:  I'll follow up with the actual signature names.

I am not sure I see this is a problem: it seems like a common pattern in type slots to cast, i.e. array number slots will use `PyArrayObject *` (must clearer) and then cast it when assigning to avoid the C-warnings because the typedef is for `PyObject *`.

Reply:
For me, the problem is it adds cognitive overhead while reading the code.m, since I need to stop and look at the definitions and compare types to see if the `void *` cast is safe. I also don't tgi k the type checking is useless, I made a few type errors that the typedefs in the C API caught while developing stringdtype.